File: televisionbingecalculator/src/main/java/com/vanillax/televisionbingecalculator/app/TBC/adapters/ShowsAdapter.java
package com.vanillax.televisionbingecalculator.app.TBC.adapters;

import android.support.annotation.NonNull;

import com.vanillax.televisionbingecalculator.app.R;
import com.vanillax.televisionbingecalculator.app.ServerAPI.TV.ShowPosterListing;
import com.vanillax.televisionbingecalculator.app.Util.BindingAdapter.BaseDataBindingAdapter;
import com.vanillax.televisionbingecalculator.app.Util.BindingAdapter.DataBoundViewHolder;
import com.vanillax.televisionbingecalculator.app.databinding.ShowCardBinding;
import com.vanillax.televisionbingecalculator.app.view.LandingActivityMain;
import com.vanillax.televisionbingecalculator.app.viewmodel.LandingActivityViewModel;
import com.vanillax.televisionbingecalculator.app.viewmodel.ShowPosterViewModelItem;

import java.util.ArrayList;
import java.util.List;

/**
 * Created by mitchross on 2/13/17.
 */

public class ShowsAdapter extends BaseDataBindingAdapter<ShowCardBinding>
{
	private List<ShowPosterViewModelItem> showsViewModelItems;
	private LandingActivityViewModel.LandingActivityViewCallback listener;



	public ShowsAdapter(  )
	{
		showsViewModelItems = new ArrayList<>(  );
	}

	public void setListener( LandingActivityViewModel.LandingActivityViewCallback listener)
	{
		this.listener = listener;
	}


	public void setShowsViewModelItems( @NonNull List<ShowPosterListing> showPosterListings, LandingActivityMain.SearchType searchType)
	{
		showsViewModelItems.clear();
		for(ShowPosterListing listing : showPosterListings)
		{
			this.showsViewModelItems.add( new ShowPosterViewModelItem( listing, searchType ) );
		}

		notifyDataSetChanged();
	}



	@Override
	protected void bindItem( DataBoundViewHolder<ShowCardBinding> holder, int position, List<Object> payloads )
	{
		holder.binding.setViewModel( showsViewModelItems.get( position ));
		holder.binding.setListener( listener );
	}

	@Override
	public int getItemCount()
	{
		return showsViewModelItems.size();
	}

	@Override
	public int getItemViewType( int position )
	{
		return R.layout.show_card;
	}
}

File: televisionbingecalculator/src/main/java/com/vanillax/televisionbingecalculator/app/kotlin/adapters/ShowsAdapter.kt
package com.vanillax.televisionbingecalculator.app.kotlin.adapters

import androidx.databinding.ViewDataBinding
import com.vanillax.televisionbingecalculator.app.R

import com.vanillax.televisionbingecalculator.app.kotlin.network.response.ShowPosterListing

import com.vanillax.televisionbingecalculator.app.kotlin.viewmodels.PosterThumbnailViewModel
import com.vanillax.televisionbingecalculator.app.util.bindingadapter.BaseDataBindingAdapter
import com.vanillax.televisionbingecalculator.app.util.bindingadapter.DataBoundViewHolder
import java.util.*

/**
 * Created by mitchross on 2/13/17.
 */

class ShowsAdapter(private val callback: (id: Int, url: String, title: String) -> Unit) : BaseDataBindingAdapter<ViewDataBinding>() {

    private val posterThumbnailViewModels: MutableList<PosterThumbnailViewModel> = arrayListOf()
    private var showPosterListings: List<ShowPosterListing> = emptyList()

    fun setShowsViewModelItems(showPosterListings: List<ShowPosterListing>) {

        if (this.showPosterListings == showPosterListings) {
            return
        }

        posterThumbnailViewModels.clear()
        for (listing in showPosterListings) {
            this.posterThumbnailViewModels.add(PosterThumbnailViewModel(listing, callback))
        }

        notifyDataSetChanged()

        this.showPosterListings = showPosterListings
    }

    override fun bindItem(holder: DataBoundViewHolder<ViewDataBinding>?, position: Int, payloads: MutableList<Any>?) {

        if (holder != null) {
            holder.binding.setVariable(BR.viewModel, posterThumbnailViewModels[position])
        }
    }

    override fun getItemCount(): Int {
        return posterThumbnailViewModels.size
    }

    override fun getItemViewType(position: Int): Int {
        return R.layout.poster_thumbnail_item
    }
}
